### PR TITLE
New Ref to Install Multiple Operators on a Cluster at One Time

### PR DESCRIPTION
**Pull request by calebevans**
## Purpose

This ref should be used to install one or more operators in a cluster.

## Process

This ref consumes a JSON blob from the [`OPERATORS` variable](#variables) and parses the JSON using the [`jq` command](https://stedolan.github.io/jq/). Using the parsed JSON, the script will iterate through the list of operators, the following happens for each operator:
1. Creates the `{install_namespace}`
   - If the namespace already exists, nothing happens.
2. Deploys a new operator group in the `{install_namespace}` with the name `{install_namespace}-operator-group`. The operator group's target namespace will be set to `{target_namespaces}`.
3. Creates a operator hub subscription in `{install_namespace}` using the `{name}`, `{channel}`, and `{source}` values provided.
4. Verifies the operator gets installed
   - Does a check every 30 seconds with 30 retries to verify the operator is installed

### Defining `OPERATORS`

The `OPERATORS` variable is a little different from other environment variables. This variable must be valid JSON to work. To define this variable correctly, please use these templates `env` blocks:

**Use with ONE operator:**
```yaml
env:
    OPERATORS: |
    [
        {"name": "", "source": "", "channel": "", "install_namespace": "", "target_namespaces": ""}
    ]
```

**Use with MULTIPLE operators:**
```yaml
env:
    OPERATORS: |
    [
        {"name": "", "source": "", "channel": "", "install_namespace": "", "target_namespaces": ""},
        {"name": "", "source": "", "channel": "", "install_namespace": "", "target_namespaces": ""}
    ]
```

#### Variable Definitions

- **`name`**: The package name of the optional operator to install. Example: `"mtr-operator"`.
- **`source`**: The catalog source name. Example: `"redhat-operators"`
- **`channel`**: The channel from which to install the package. This value can be set to `"!default"` if you'd like to always install from the default channel. Example: `"release-2.7"`.
- **`install_namespace`**: The namespace into which the operator and catalog will be installed. Example: `"mtr-namespace"`.
- **`target_namespaces`**: A comma-separated list of namespaces the operator will target. This value can be set to `"!install"` to use the `install_namespace` value. Example: `"mtr,ocm"`

### Infrastructure

- A provisioned test cluster to target.

### Variables

- `OPERATORS` 
  - **Definition**: A JSON list of operators to install. Please see the [Defining `OPERATORS`](#defining-operators) section of this document for more information.
  - **If left empty**: The script will fail.

### CODE REVIEW

File: ci-operator/step-registry/lp-interop-tooling/install-operators/lp-interop-tooling-install-operators-ref.yaml
Comment (calebevans) [reply ci-operator/step-registry/lp-interop-tooling/install-operators/lp-interop-tooling-install-operators-ref.yaml:2]: Yeah, I think that works for me! I will update in an upcoming commit.

File: ci-operator/step-registry/lp-interop-tooling/OWNERS
Comment (calebevans) [reply ci-operator/step-registry/lp-interop-tooling/OWNERS]: Resolved in [**16b5500**037ffee6371972b08168f7d81b77b9026](https://github.com/openshift/release/pull/36381/commits/16b5500037ffee6371972b08168f7d81b77b9026)

File: ci-operator/step-registry/lp-interop-tooling/install-operators/lp-interop-tooling-install-operators-ref.yaml
Comment (calebevans) [reply ci-operator/step-registry/lp-interop-tooling/install-operators/lp-interop-tooling-install-operators-ref.yaml]: Resolved in [**16b5500**037ffee6371972b08168f7d81b77b9026](https://github.com/openshift/release/pull/36381/commits/16b5500037ffee6371972b08168f7d81b77b9026)

File: ci-operator/step-registry/install-operators/install-operators-ref.yaml
Comment (ryankwilliams) [inline ci-operator/step-registry/install-operators/install-operators-ref.yaml:3]: I feel if this is going to be a step that is global to everyone (available in the step registry), then as an end user. All they should need to do is include this new step within their job config (e.g. `- ref: install-operators`) along with an image already available in the cluster. Any additional steps such as creating the container image to use for the step should be avoided.

As taking ownership of this new step, it would make sense to contribute a new base image available to all OCP clusters within prow. This way the user can just define the base image over needing to do a `dockerfile_literal` within every job config using this step.

Based on my knowledge, we would need to do the following to have a new base image be made available to the clusters:

1. Create a new repository for this step with the Dockerfile for building the image (probably repo in the CSPI-QE org)
2. Wire up a job to automatically build the image and push the image to a container image registry for every PR merge
3. Open PR to release repository to include this new image to be mirrored into the clusters making it available for use.

Example how `osde2e` mirrors their image into the clusters: https://github.com/openshift/release/blob/master/core-services/image-mirroring/supplemental-ci-images/mapping_supplemental_ci_images_ci#L27

https://docs.ci.openshift.org/docs/how-tos/external-images/#mirror-public-images

Then from the job config they just need to declare it within the base images:

```
base_images:
  osde2e:
    name: osde2e
    namespace: ci
    tag: latest
```

So we could have a base image for this step such as:

```
base_images:
  install-operators:
    name: install-operators
    namespace: ci
    tag: latest
```

Comment (calebevans) [reply ci-operator/step-registry/install-operators/install-operators-ref.yaml:3]: I like that, I wasn't aware we had any control over which images get put into the image registry. 

@ascerra are you okay with this approach too?

Comment (ryankwilliams) [reply ci-operator/step-registry/install-operators/install-operators-ref.yaml:3]: @calebevans Since this is quite a simple container image, the repo created could even benefit from using the [GH container registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry). I've used it before, works nicely and can tie it into a GH action.

Comment (calebevans) [reply ci-operator/step-registry/install-operators/install-operators-ref.yaml:3]: @ryankwilliams yeah I can give that a shot. Do you think this should be its own repo though? Maybe we can just do a lp-interop-images or cspi-images repo for this sort of thing?

File: ci-operator/step-registry/install-operators/README.md
Comment (ryankwilliams) [inline ci-operator/step-registry/install-operators/README.md:13]: We can probably remove this section right?

File: ci-operator/step-registry/install-operators/README.md
Comment (ryankwilliams) [inline ci-operator/step-registry/install-operators/README.md:75]:
```suggestion
This ref requires `jq` and `oc` to execute correctly. We have created an image in the [CSPI-QE/images](https://github.com/CSPI-QE/images) repository that get's pushed to Quay. The image is then mirrored into OpenShift for use in this step.
```

Comment (madunn) [reply ci-operator/step-registry/install-operators/README.md:75]: > 

```suggestion
This ref requires `jq` and `oc` to execute correctly. We have created an in the [CSPI-QE/images]() repository that gets pushed to Quay. The image is then mirrored into OpenShift for use in this step.
```

File: ci-operator/step-registry/install-operators/README.md
Comment (madunn) [inline ci-operator/step-registry/install-operators/README.md:25]:
```suggestion
3. Creates an operator hub subscription in `{install_namespace}` using the `{name}`, `{channel}`, and `{source}` values provided.
```

File: ci-operator/step-registry/install-operators/install-operators-commands.sh
Comment (madunn) [inline ci-operator/step-registry/install-operators/install-operators-commands.sh:7]:
```suggestion
# Read each operator in the JSON provided to an item in a BASH array.
```

**Review by ascerra — APPROVED** [review]

File: ci-operator/step-registry/install-operators/install-operators-ref.yaml
Comment (ascerra) [inline ci-operator/step-registry/install-operators/install-operators-ref.yaml:3]: I found an imagestream named cli-jq

I'm wondering if this will satisfy this need rather than having us maintain this
https://console-openshift-console.apps.ci.l2s4.p1.openshiftapps.com/k8s/ns/ocp/imagestreamtags/cli-jq%3Alatest

It seem like we could just use
```
from_image:
    namespace: ocp
    name: cli-jq
    tag: cli
```

Comment (calebevans) [reply ci-operator/step-registry/install-operators/install-operators-ref.yaml:3]: I'd rather do that. Let me give it a shot real quick.

Comment (calebevans) [reply ci-operator/step-registry/install-operators/install-operators-ref.yaml:3]: @ascerra that worked perfectly. I will remove the work we have done from this PR. I'll keep that repo around for now though. 

CC: @ryankwilliams

**Comment by danilo-gemoli** [comment]
@ascerra I can put an `lgtm` to move your PR forward whenever you feel ready but I have usually nothing to say regarding users' tests

File: ci-operator/step-registry/install-operators/OWNERS
Comment (ryankwilliams) [inline ci-operator/step-registry/install-operators/OWNERS:1]: Before we get too far into PR's being opened/merged, lets solve the approves/reviewers now (my personal opinion).

We just need to reach out to prob `forum-dp-platform` and get admin of the `openshift` org to create a new team and provide admin access to certain people. Then those admins can control who is added.

There are a couple approaches we can go here regarding teams created:

- **Very Simple**: One Team (e.g. `cspi-maintainers`) which would can be used for the `approvers` and `reviewers` roles.
- **Somewhat Simple**: Two teams (e.g. `cspi-approvers` & `cspi-reviewers`) which then would be assigned to their correct roles `approvers || reviewers`. These teams then can apply to any type of jobs (e.g. layered product interop testing, service integration testing, etc). Everyone is in the loop.
- **Starting To Get Complicated**: Two teams (e.g. `cspi-lp-maintainers` && `cspi-<X>-maintainers`) which then would have people in different teams based on projects. Then each team would be part of the `approvers` and `reviewers`.
- **Lets Get Complicated!**: Four teams (e.g. `cspi-lp-approvers`, `cspi-lp-reviewers`, `cspi-<X>-approvers`, `cspi-<X>-reviewers`. Based on above just being more granular around who can do what to a PR.

This can kick off the discussion prior to submitting  any request for team(s). Doing this now just will reduce the amount of future PR's needed to change everyone's users if needed.

**Review by ryankwilliams — CHANGES_REQUESTED** [review]

File: ci-operator/step-registry/install-operators/OWNERS
Comment (ryankwilliams) [reply ci-operator/step-registry/install-operators/OWNERS:1]: Since this is a more of a nice to have and not a blocker, we can ignore this for now. After some further research, using teams within OWNER files is not supported by prow. There is other ways to create aliases that can be used. This can be explored at a later time.

https://docs.prow.k8s.io/docs/components/plugins/approve/approvers/

**Review by ryankwilliams — APPROVED** [review]

**Pull request merged**
